File: lib/boilerpl8.rb
# -*- coding: utf-8 -*-

###
### $Release: 0.2.0 $
### $Copyright: copyright(c) 2016 kuwata-lab.com all rights reserved $
### $License: MIT License $
###

require 'open-uri'
require 'json'
require 'fileutils'


module Boilerpl8


  RELEASE = '$Release: 0.2.0 $'.split()[1]


  module ShellHelper
    PROMPT = "[boilerpl8]$"

    def _prompt()
      return PROMPT
    end

    def rm_rf(path)
      puts "#{_prompt()} rm -rf #{path}"
      FileUtils.rm_rf path
    end

    def mv(oldpath, newpath)
      puts "#{_prompt()} mv #{oldpath} #{newpath}"
      File.rename oldpath, newpath
    end

    def sys(command)
      puts "#{_prompt()} #{command}"
      system command
    end

    def chdir(dir, &block)
      puts "#{_prompt()} cd #{dir}"
      Dir.chdir(dir, &block)
      puts "#{_prompt()} cd -"
    end

  end


  INITIALIZER_SCRIPTS = [
    ["__init.rb"   , "ruby"  ],
    ["__init.py"   , "python"],
    ["__init.js"   , "node"  ],
    ["__init.pl"   , "perl"  ],
    ["__init.php"  , "php"   ],
    ["__init.lua"  , "lua"   ],
    ["__init.exs"  , "elixir"],
    ["__init.sh"   , "bash"  ],
  ]


  class Operation
    include ShellHelper

    def do_everything(boilerplate_name, target_dir, options)
      url, filename = resolve(boilerplate_name, options)
      filepath = download(url, filename)
      basedir = extract(filepath, target_dir)
      ok = kick_initializer(basedir)
      return ok    # true or false
    end

    protected

    def resolve(arg, options)
      raise NotImplementedError.new("#{self.class.name}#resolve(): not implemented yet.")
    end

    def download(url, filename)
      print "Download from #{url} ..."
      content = open(url) {|f| f.read }
      puts " done."
      File.open(filename, 'wb') {|f| f.write(content) }
      return filename
    end

    def extract(filename, basedir)
      filename =~ /\.(zip|tgz|tar\.(gz|bz2|xz))\z/  or
        err("#{filename}: expected '*.zip' or '*.tar.gz'")
      base = File.basename($`)
      basedir ||= base
      rm_rf basedir if File.exist?(basedir)
      #
      case filename
      when /\.zip\z/
        tmpdir = basedir + ".tmp"
        sys "unzip -q -d #{tmpdir} #{filename}"
        paths = Dir.glob("#{tmpdir}/*")
        if paths.length == 1 && File.directory?(paths[0])
          mv paths[0], basedir
          rm_rf tmpdir
        else
          mv tmpdir, basedir
        end
      else
        sys "tar xf #{filename}"
        mv base, basedir if base != basedir
      end
      #
      return basedir
    end

    def kick_initializer(basedir)
      chdir(basedir) do
        INITIALIZER_SCRIPTS.each do |script, lang|
          return sys "#{lang} #{script}" if File.exist?(script)
        end
      end
      return true   # initializer script not exist
    end

    def err(msg)
      raise CommandOptionError.new(msg)
    end

    def self.err(msg)
      raise CommandOptionError.new(msg)
    end

    public

    ALL = []

    def self.inherited(klass)
      ALL << klass
    end

    def self.create(boilerplate_name)
      #; [!xr4c6] reports error when argument has no schema.
      boilerplate_name =~ /\A(\w+:)/  or
        err("#{boilerplate_name}: expected 'github:' or 'file:' schema.")
      schema = $1
      #; [!95h3f] reports error when argument has unknown schema.
      klass = ALL.find {|cls| cls.const_get(:SCHEMA) == schema }  or
        err("#{boilerplate_name}: unknown schema (expected 'github:' or 'file:').")
      return klass.new()
    end

  end


  class FileSystemOp < Operation

    SCHEMA = "file:"

    protected

    def resolve(arg, options)
      arg =~ %r'\Afile:(.+)'  or err("#{arg}: unexpected format.")
      filepath = $1
      return filepath, File.basename(filepath)
    end

    def download(filepath, filename)
      return filepath
    end

  end


  class GithubOp < Operation

    SCHEMA = "github:"

    protected

    def resolve(arg, options)
      arg =~ %r'\Agithub:([^/]+)/([^/]+)\z'  or err("#{arg}: unexpected format.")
      user, repo = $1, $2
      #
      suffix = options['B'] ? "" : "-boilerpl8"
      api_url = "https://api.github.com/repos/#{user}/#{repo}#{suffix}/releases"
      begin
        json_str = open(api_url) {|f| f.read }
      rescue OpenURI::HTTPError => ex
        hint = options['B'] \
             ? "confirm repository name, or try without '-B' option." \
             : "confirm repository name, or maybe you missed '-B' option."
        err("#{arg}: repository not found\n  (api: GET #{api_url})\n  (Hint: #{hint})")
      end
      #
      json_arr = JSON.parse(json_str)
      dict = json_arr[0]
      asset = (dict["assets"] || [])[0]
      if asset
        zip_url = asset["browser_download_url"]
        filename = zip_url ? File.basename(zip_url) : nil
      else
        zip_url = dict["zipball_url"]
        filename = "#{repo}_#{dict['tag_name']}.zip"
      end
      zip_url  or
        err("ERROR: can't find zip file under github.com/#{user}/#{repo}/releases")
      return zip_url, filename
    end

  end


  class MainApp

    COMMAND_OPTIONS = [
      "-h, --help       :  help",
      "-v, --version    :  version",
      "-B               :  not append '-boilerpl8' to github repo name",
    ]

    def self.main
      begin
        status = self.new.run(*ARGV)
        exit status
      rescue CommandOptionError => ex
        $stderr.puts ex.message
        exit 1
      end
    end

    def initialize(script_name=nil)
      @script_name = script_name || File.basename($0)
    end

    def run(*args)
      parser = CommandOptionParser.new(COMMAND_OPTIONS)
      options = parser.parse(args)
      #
      if options['help']
        puts help_message()
        return 0
      end
      #
      if options['version']
        puts RELEASE
        return 0
      end
      #
      boilerplate_name = args[0]  # ex: "github:kwatch/hello-ruby"
      target_dir       = args[1]  # ex: "mygem1"
      #; [!eqisx] reports error when boilerplate name or target dir is not specified.
      boilerplate_name  or raise err("#{@script_name}: argument required.")
      target_dir        or raise err("#{@script_name}: target directory name required.")
      #
      op = Operation.create(boilerplate_name)
      ok = op.do_everything(boilerplate_name, target_dir, options)
      return ok ? 0 : 1
    end

    def help_message
      script = @script_name
      buf = <<"END"
#{script} -- download boilerplate files

Usage:
  #{script} [options] github:<USER>/<REPO> <DIR>
  #{script} [options] file:<PATH> <DIR>

Options:
END
      COMMAND_OPTIONS.each {|s| buf << "  #{s}\n" }
      buf << <<"END"

Examples:

  ## download boilerplate files from github
  $ #{script} github:kwatch/hello-ruby mygem1             # for ruby
  $ #{script} github:kwatch/hello-python mypkg1           # for python
  $ #{script} github:kwatch/keight-ruby myapp1            # for keight.rb

  ## '-B' option doesn't append '-boilerpl8' to github repo name
  $ #{script} -B github:h5bp/html5-boilerplate website1   # for html5

  ## expand boilerplate files
  $ #{script} file:./keight-ruby.tar.gz myapp1

END
    end

    private

    def err(msg)
      raise CommandOptionError.new(msg)
    end

  end


  class CommandOptionError < StandardError
  end


  CommandOptionDefinition = Struct.new(:short, :long, :param, :desc)


  class CommandOptionParser

    def initialize(optdef_strs)
      @optdef_strs = optdef_strs
      @optdefs = []
      optdef_strs.each do |optdef_str|
        case optdef_str
        when /-(\w), --(\w[-\w]*)(?:=(\S+))?\s*:\s*(\S.*)?/ ; t = [$1, $2, $3, $4]
        when /-(\w)(?:\s+(\S+))?\s*:\s*(\S.*)?/             ; t = [$1, nil, $2, $3]
        when /--(\w[-\w]*)(?:=(\S+))?\s*:\s*(\S.*)?/        ; t = [nil, $1, $2, $3]
        else
          raise "unexpected option definition: #{optdef_str}"
        end
        short, long, param, desc = t
        @optdefs << CommandOptionDefinition.new(short, long, param, desc)
      end
    end

    attr_reader :optdefs

    def parse(args)
      options = {}
      while ! args.empty? && args[0].start_with?('-')
        argstr = args.shift
        if argstr.start_with?('--')
          rexp = /\A--([-\w]+)(?:=(.*))?\z/
          argstr =~ rexp                   or err("#{argstr}: invalid option format.")
          name, value = $1, $2
          optdef = find_by(:long, name)    or err("--#{name}: unknown option.")
          optdef.param.nil? || value       or err("#{argstr}: argument required.")
          optdef.param      || value.nil?  or err("#{argstr}: unexpected argument.")
          options[optdef.long] = value || true
        else
          n = argstr.length
          i = 0
          while (i += 1) < n
            ch = argstr[i]
            optdef = find_by(:short, ch)   or err("-#{ch}: unknown option.")
            if optdef.param.nil?   # no arguments
              options[optdef.long || optdef.short] = true
            else                   # argument required
              param = argstr[(i+1)..-1]
              param = args.shift if param.empty?
              param  or err("-#{ch}: argument required.")
              options[optdef.long || optdef.short] = param
              break
            end
          end
        end
      end
      return options
    end

    private

    def find_by(key, value)
      return @optdefs.find {|x| x.__send__(key) == value }
    end

    def err(msg)
      raise CommandOptionError.new(msg)
    end

  end


end


if __FILE__ == $0
  Boilerpl8::MainApp.main()
end

File: boilerpl8.py
# -*- coding: utf-8 -*-

###
### $Release: 0.0.0 $
### $Copyright: copyright(c) 2016 kuwata-lab.com all rights reserved $
### $License: MIT License $
###

import sys, os, re
PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3
assert PY2 or PY3

import json
import shutil
from glob import glob
if PY2:
    from urllib2 import urlopen
elif PY3:
    from urllib.request import urlopen


__release__ = '$Release: 0.0.0 $'.split()[1]


class Operation(object):

    def resolve(self, arg, options):
        raise NotImplementedError("%s.resolve(): not implemented yet." % self.__class__.__name__)

    def download(self, url, filename):
        sys.stdout.write("Download from %s ..." % url)
        resp = urlopen(url)
        content = resp.read()
        sys.stdout.write(" done.\n")
        with open(filename, 'wb') as f:
            f.write(content)
        return filename

    def extract(self, filename, basedir):
        m = re.search(r'\.(zip|tgz|tar\.(gz|bz2|xz))$', filename)
        if not m:
            raise _err("%s: expected '*.zip' or '*.tar.gz'" % filename)
        base = filename[:m.start(0)]
        basedir = basedir or base
        if os.path.exists(basedir):
            print("$ rm -rf %s" % basedir)
            shutil.rmtree(basedir)
        #
        if filename.endswith('.zip'):
            print("$ unzip -q -d %s.tmp %s" % (basedir, filename))
            os.system("unzip -q -d %s.tmp %s" % (basedir, filename))
            paths = glob("%s.tmp/*" % basedir)
            if len(paths) == 1 and os.path.isdir(paths[0]):
                print("$ mv %s %s" % (paths[0], basedir))
                os.rename(paths[0], basedir)
                print("$ rm -rf %s.tmp" % basedir)
                shutil.rmtree("%s.tmp" % basedir)
            else:
                print("$ mv %s.tmp %s" % (basedir, basedir))
                os.rename("%s.tmp" % basedir, basedir)
        else:
            print("$ tar xf %s" % filename)
            os.system("tar xf %s" % filename)
            if base != basedir:
                print("$ mv %s %s" % (base, basedir))
                os.rename(base, basedir)
        #
        return basedir

    def kick_initializer(self, basedir):
        print("$ cd %s" % basedir)
        os.chdir(basedir)
        try:
            for script, lang in INITIALIZER_SCRIPTS:
                if os.path.isfile(script):
                    print("$ %s %s" % (lang, script))
                    os.system(" %s %s" % (lang, script))
                break
        finally:
            print("$ cd -")
            os.chdir("..")

    ALL = {}

    @classmethod
    def create(cls, *args):
        m = re.match(r'^(\w+:)', args[0])
        if not m:
            raise _err("%s: expected 'github:' or 'file:' schema." % args[0])
        schema = m.group(1)
        klass = cls.ALL.get(schema)
        if klass is None:
            raise _err("%s: unknown schema." % args[0])
        return klass()


class FileSystemOperation(Operation):

    SCHEMA = "file:"

    def resolve(self, arg, options):
        m = re.match(r'^file:(.+)', arg)
        if not m:
            raise _err("%s: unexpected format." % arg)
        filepath = m.group(1)
        return filepath, os.path.basename(filepath)

    def download(self, filepath, filename):
        return filepath


class GithubOperation(Operation):

    SCHEMA = "github:"

    def resolve(self, arg, options):
        m = re.match(r'^github:([^/]+)/([^/]+)$', arg)
        if not m:
            raise _err("%s: unexpected format." % arg)
        user, repo = m.groups()
        #
        suffix = ("" if options.get('B') else "-boilerpl8")
        api_url = "https://api.github.com/repos/%s/%s%s/releases" % (user, repo, suffix)
        try:
            resp = urlopen(api_url)
            json_str = resp.read()
        except Exception as ex:
            if options.get('B'):
                hint = "confirm repository name, or try without '-B' option."
            else:
                hint = "confirm repository name, or maybe you missed '-B' option."
            raise _err("%s: repository not found.\n" % repo +
                       "  (api: GET %s)\n" % api_url +
                       "  (Hint: %s)" % hint)
        #
        if isinstance(json_str, bytes):
            json_str = json_str.decode('utf-8')
        json_arr = json.loads(json_str)
        d = json_arr[0]
        asset = (d["assets"][0] if d["assets"] else None)
        if asset:
            zip_url = asset["browser_download_url"]
            filename = (os.path.basename(zip_url) if zip_url else None)
        else:
            zip_url = d["zipball_url"]
            filename = "%s_%s.zip" % (repo, d['tag_name'])
        if not zip_url:
            raise _err("ERROR: can't find zip file under github.com/%s/%s/releases" % (user, repo))
        return zip_url, filename


for cls in (FileSystemOperation, GithubOperation):
    Operation.ALL[cls.SCHEMA] = cls


INITIALIZER_SCRIPTS = [
    ("__init.rb"   , "ruby"  ),
    ("__init.py"   , "python"),
    ("__init.js"   , "node"  ),
    ("__init.pl"   , "perl"  ),
    ("__init.php"  , "php"   ),
    ("__init.lua"  , "lua"   ),
    ("__init.exs"  , "elixir"),
    ("__init.sh"   , "bash"  ),
]


class App(object):

    COMMAND_OPTIONS = [
        "-h, --help       :  help",
        "-v, --version    :  version",
        "-B               :  not append '-boilerpl8' to github repo name",
    ]

    def __init__(self, script_name):
        self.script_name = script_name

    def run(self, *args):
        args = list(args)
        parser = CommandOptionParser(self.COMMAND_OPTIONS)
        options = parser.parse(args)
        #
        if options.get('help'):
            print(self.help_message())
            return 0
        #
        if options.get('version'):
            print(__release__)
            return 0
        #
        if not args:
            raise _err("%s: argument required." % self.script_name)
        op = Operation.create(*args)
        url, filename = op.resolve(args[0], options)
        filepath = op.download(url, filename)
        basedir = op.extract(filepath, (args[1] if len(args) >= 2 else None))
        op.kick_initializer(basedir)
        return 0

    def help_message(self):
        buf = []; add = buf.append
        add(r"""
{script} -- download boilerplate files

Usage:
  {script} [options] github:<USER>/<REPO> <DIR>
  {script} [options] file:<PATH> <DIR>

Options:
"""[1:])
        for s in self.COMMAND_OPTIONS:
            add("  %s\n" % s)
        add(r"""
Examples:

  ## download boilerplate files from github
  $ {script} github:kwatch/hello-python mypkg1           # for python
  $ {script} github:kwatch/hello-ruby mygem1             # for ruby
  $ {script} github:kwatch/keight-python myapp1          # for keight.py

  ## '-B' option doesn't append '-boilerpl8' to github repo name
  $ {script} -B github:h5bp/html5-boilerplate website1   # for html5

  ## expand boilerplate files
  $ {script} file:./keight-python.tar.gz myapp1
""")
        return "".join(buf).format(script=self.script_name)



class CommandOptionError(Exception):
    pass


def _err(msg):
    raise CommandOptionError(msg)



class CommandOptionDefinition(object):

    def __init__(self, short, long, param, desc):
        self.short = short
        self.long  = long
        self.param = param
        self.desc  = desc


class CommandOptionParser(object):

    def __init__(self, optdef_strs):
        def fn(s):
            short = long = param = desc = None
            m = re.match(r'^-(\w), --(\w[-\w]*)(?:=(\S+))?\s*:\s*(\S.*)?$', s)
            if m:
                short, long, param, desc = m.groups()
                return short, long, param, desc
            m = re.match(r'^-(\w)(?:\s+(\S+))?\s*:\s*(\S.*)?$', s)
            if m:
                short, params, desc = m.groups()
                return short, long, param, desc
            m = re.match(r'^--(\w[-\w]*)(?:=(\S+))?\s*:\s*(\S.*)?$', s)
            if m:
                long, params, desc = m.groups()
                return short, long, param, desc
            return None
        #
        self.optdef_strs = optdef_strs
        self.optdefs = []
        for optdef_str in optdef_strs:
            t = fn(optdef_str.strip())
            if t is None:
                raise TypeError("unexpected option definition: %s" % optdef_str)
            short, long, param, desc = t
            self.optdefs.append(CommandOptionDefinition(short, long, param, desc))

    def parse(self, args):
        assert isinstance(args, list)
        options = {}
        while args and args[0].startswith('-'):
            argstr = args.pop(0)
            if argstr.startswith('--'):
                m = re.match(r'^--([-\w]+)(?:=(.*))?$', argstr)
                if not m:
                    raise _err("%s: invalid option format.")
                name, value = m.groups()
                optdef = self._find_by('long', name)
                if optdef is None:
                    raise _err("--%s: unknown option." % name)
                if optdef.param is not None and value is None:
                    raise _err("%s: argument required." % argstr)
                if optdef.param is None and value is not None:
                    raise _err("%s: unexpected argument." % argstr)
                options[optdef.long] = (True if value is None else vlue)
            else:
                n = len(argstr)
                i = 1
                while i < n:
                    ch = argstr[i]
                    optdef = self._find_by('short', ch)
                    if optdef is None:
                        raise _err("-%s: unknown option." % ch)
                    if optdef.param is None:   # no arguments
                        options[optdef.long or optdef.short] = True
                        i += 1
                    else:                      # argument required
                        param = argstr[(i+1):]
                        if not param:
                            if args:
                                param = args.pop(0)
                            else:
                                raise _err("-%s: argument required." % ch)
                        options[optdef.long or optdef.short] = param
                        break
        return options

    def _find_by(self, key, value):
        for x in self.optdefs:
            if getattr(x, key, None) == value:
                return x
        return None


def main():
    script_name = os.path.basename(sys.argv[0])
    args = sys.argv[1:]
    try:
        status = App(script_name).run(*args)
    except CommandOptionError as ex:
        sys.stderr.write("%s\n" % (ex,))
        status = 1
    sys.exit(status)


if __name__ == '__main__':
    main()